validator/src/main.rs

```diff
@@ -0,0 +1,249 @@
+    outbound_proxy_addr: SocketAddr,
+
+    #[clap(long)]
+    target_addr: SocketAddr,
+
+    #[clap(parse(try_from_str = parse_timeout), long, default_value = "120s")]
+    timeout: std::time::Duration,
+}
+
+// ERRNO 95: Operation not supported
+const UNSUCCESSFUL_EXIT_CODE: i32 = 95;
+
+#[tokio::main(flavor = "current_thread")]
+async fn main() -> Result<()> {
+    let Args {
+        log_level,
+        log_format,
+        outbound_proxy_addr,
+        target_addr,
+        timeout,
+    } = Args::parse();
+
+    log_format.try_init(log_level)?;
+
+    let rng_resp = Alphanumeric.sample_string(&mut rand::thread_rng(), 8);
+    info!(%outbound_proxy_addr, %target_addr, "Validating outbound traffic is redirected to proxy's outbound port");
+
+    let (shutdown_tx, shutdown_rx) = kubert::shutdown::sigint_or_sigterm()?;
+    let notify_ready = Arc::new(Notify::new());
+
+    let server_task = tokio::spawn(outbound_serve(
+        outbound_proxy_addr,
+        notify_ready.clone(),
+        shutdown_rx,
+        rng_resp.clone(),
+    ));
+
+    let validation_task = tokio::spawn(validate_outbound_redirect(
+        target_addr,
+        notify_ready,
+        timeout,
+        rng_resp,
+    ));
+
+    tokio::select! {
+        task = validation_task => {
+            if let Err(error) = task.expect("Failed to run validator task") {
+                error!(%error, "Failed to validate");
+                exit(UNSUCCESSFUL_EXIT_CODE);
+            }
+
+             info!("Validation passed successfully...exiting");
+            Ok(())
+        }
+
+         _ = shutdown_tx.signaled() => {
+            if let Err(error) = server_task.await.expect("Failed to run outbound server task") {
+                error!(%error, "Failed to validate outbound routing configuration");
+            } else {
+                error!("Failed to validate due to server terminating early");
+            }
+
+             exit(UNSUCCESSFUL_EXIT_CODE);
+        }
+    }
+}
+
+#[tracing::instrument(level = "debug", skip_all)]
+async fn validate_outbound_redirect(
+    target_addr: SocketAddr,
+    ready: Arc<Notify>,
+    timeout: time::Duration,
+    expected_resp: String,
+) -> Result<()> {
+    if tokio::time::timeout(timeout, ready.notified())
+        .await
+        .is_err()
+    {
+        bail!("timed-out ({:?}) waiting for server to be ready", timeout);
+    }
+
+    let mut stream = {
+        debug!("Building validation client");
+        let socket = TcpStream::connect(target_addr).await?;
+        debug_assert_eq!(target_addr, socket.peer_addr().unwrap());
+        debug!("Connection established");
+        socket
+    };
+
+    tokio::select! {
+        is_readable = stream.readable() => {
+            is_readable.context("cannot read off client socket")?;
+        }
+
+         () = tokio::time::sleep(timeout) => {
+            bail!("timed out ({:?}) waiting for socket to become readable", timeout);
+        }
+    };
```

Comment:
Yeah, I agree. I think your suggestion of using a reserved IP address works much better here. At first, I was using the k8s address. When the firewall wasn't configured, we'd establish a connection almost instantly, but then we'd timeout on the stream being readable (since the k8s api server didn't send anything to us).

Like you said it's odd the way it works, makes more sense to have a single timeout to cover everything in totality, and also it makes sense to have a smaller value for that timeout 👍🏻